### ISSUE TITLE
invalid size error

### ISSUE THREAD

**Issue by ptorrent** [xeokit/xeokit-convert#72]
Hello,

I'm trying this command:

./node-v16.17.0-linux-x64/bin/node --max-old-space-size=25192 ./node_modules/@xeokit/xeokit-convert/convert2xkt.js -s test.ifc -o test.ifc.xkt -l

The ifc file size is 750M

Everything is looking good but after a while this is the error I got :

" JavaScript invalid size error"

Any idea

```
[convert2xkt] Input file parsed OK. Building XKT document...
[convert2xkt] XKT document built OK. Writing to XKT file...


#
# Fatal error in , line 0
# Fatal JavaScript invalid size error 169220804
#
#
#
#FailureMessage Object: 0x7fffc2b0caa0
 1: 0xb6ca81  [./node-v16.17.0-linux-x64/bin/node]
 2: 0x1bef1a4 V8_Fatal(char const*, ...) [./node-v16.17.0-linux-x64/bin/node]
 3: 0xe63268  [./node-v16.17.0-linux-x64/bin/node]
 4: 0x1010b12  [./node-v16.17.0-linux-x64/bin/node]
 5: 0x1011456  [./node-v16.17.0-linux-x64/bin/node]
 6: 0x11d1b73 v8::internal::Runtime_GrowArrayElements(int, unsigned long*, v8::internal::Isolate*) [./node-v16.17.0-linux-x64/bin/node]
 7: 0x15d5439  [./node-v16.17.0-linux-x64/bin/node]
Trace/breakpoint trap

```

**Comment by ptorrent** [comment]
Looks the same with an other file :

635M

```
[convert2xkt] Input file parsed OK. Building XKT document...
[convert2xkt] XKT document built OK. Writing to XKT file...


#
# Fatal error in , line 0
# Fatal JavaScript invalid size error 169220804
#
#
#
#FailureMessage Object: 0x7fffe9224900
 1: 0xb6ca81  [./node-v16.17.0-linux-x64/bin/node]
 2: 0x1bef1a4 V8_Fatal(char const*, ...) [./node-v16.17.0-linux-x64/bin/node]
 3: 0xe63268  [./node-v16.17.0-linux-x64/bin/node]
 4: 0x1010b12  [./node-v16.17.0-linux-x64/bin/node]
 5: 0x1011456  [./node-v16.17.0-linux-x64/bin/node]
 6: 0x11d1b73 v8::internal::Runtime_GrowArrayElements(int, unsigned long*, v8::internal::Isolate*) [./node-v16.17.0-linux-x64/bin/node]
 7: 0x15d5439  [./node-v16.17.0-linux-x64/bin/node]
Trace/breakpoint trap
```

**Comment by ptorrent** [comment]
I can send you the file if needed... Will try an other file that was working with: 

https://github.com/xeokit/xeokit-xkt-utils

**Comment by ptorrent** [comment]
Bad news :( 

Was working with xeokit-xkt-utils

But not with this tool

```
[convert2xkt] Input file parsed OK. Building XKT document...
[convert2xkt] XKT document built OK. Writing to XKT file...


#
# Fatal error in , line 0
# Fatal JavaScript invalid size error 169220804
#
#
#
#FailureMessage Object: 0x7ffe1f024640
 1: 0xb6ca81  [./node-v16.17.0-linux-x64/bin/node]
 2: 0x1bef1a4 V8_Fatal(char const*, ...) [./node-v16.17.0-linux-x64/bin/node]
 3: 0xe63268  [./node-v16.17.0-linux-x64/bin/node]
 4: 0x1010b12  [./node-v16.17.0-linux-x64/bin/node]
 5: 0x1011456  [./node-v16.17.0-linux-x64/bin/node]
 6: 0x11d1b73 v8::internal::Runtime_GrowArrayElements(int, unsigned long*, v8::internal::Isolate*) [./node-v16.17.0-linux-x64/bin/node]
 7: 0x15d5439  [./node-v16.17.0-linux-x64/bin/node]
Trace/breakpoint trap

```

**Comment by ptorrent** [comment]
I note that all file failed with same error:

invalid size error **169220804**

**Comment by xeolabs** [comment]
This is not a bug with ````convert2xkt````, rather it only happens when the IFC file is too big to convert, and nodejs runs out of memory. 

````xeokit-utils``` is much simpler than ````convert2xkt````, and demands less nodejs resources, so that's likely the reason your IFC converted OK there.

You could try our other conversion pipelines, which convert to intermediate .glb and JSON files for consumption by convert2xkt, which gives better memory performance. 

Other than that, there's not much we can do about this at this stage. All you have to do to overload nodejs is give it a string in excess of (or approaching) 1GB to parse and it crashes - we don't have much control over that. Converting .glb and JSON from ifc2gltf or ifcConvert, instead of embedded glTF (which creates huge strings in convert2xkt),  is the best option so far.

**Comment by xeolabs** [comment]
Nodejs has a hard text string length limit of 1GB, which is the reason why some files fail to convert.  This is a nodejs limit, not xeokit's fault. 

We may be able to improve things by  changing the metadata format within the XKT.

Internally, convert2xkt puts all the metadata JSON in a string, and parses it from there. 

I may be able to compress the metadata format (resulting in XKT v10.1 and updates to xeokit-sdk and xeokit-convert) so that it uses perhaps ~30-40% less memory. 

The metadata uses many copies of the same long string IDs, to connect metaobjects into parent-child trees, ie:

{ id: "567dghfjfjhgfkhgfkhjgkhjghjgkjhgkjhgj", parentId: "hjkgfkjhgkjhuyweuyuyg", ...},
{ id: "432dghfjfjhgfkhgfkhjgkhjghjgkjhgkjhgj", parentId: "567dghfjfjhgfkhgfkhjgkhjghjgkjhgkjhgj", ...},
{ id: "008dghfjfjhgfkhgfkhjgkhjghjgkjhgkjhgj", parentId: "567dghfjfjhgfkhgfkhjgkhjghjgkjhgkjhgj", ...},

We can make it more compact by using "internal IDs" for the parent-child connectivity within the metadata JSON:

{ id: "567dghfjfjhgfkhgfkhjgkhjghjgkjhgkjhgj", internalId: 0,  parentId: null, ...},
{ id: "432dghfjfjhgfkhgfkhjgkhjghjgkjhgkjhgj", internalId: 1, parentId: 0, ...},
{ id: "008dghfjfjhgfkhgfkhjgkhjghjgkjhgkjhgj", internalId: 2 parentId: 0, ...},

 I will also try splitting the metadata within the XKT into several JSON chunks, and attempt to parse them separately.

**Comment by ptorrent** [comment]
Hello ! Thanks for your investigations. I think to chunks the JSON is a good idea !

Thanks for your support

**Comment by xeolabs** [comment]
I just published an optimization in xeokit-convert 1.1.3 that might help with this issue - see https://github.com/xeokit/xeokit-convert/pull/78
